lib/archive/unpack_unix.go

```
@@ -0,0 +1,198 @@
+
+		op.Debugf("processing tar header: asset(%s), size(%d)", header.Name, header.Size)
+		// skip excluded elements unless explicitly included
+		if filter.Excludes(op, header.Name) {
+			continue
+		}
+
+		// fix up path
+		stripped := strings.TrimPrefix(header.Name, filter.StripPath)
+		rebased := filepath.Join(filter.RebasePath, stripped)
+		absPath := filepath.Join(root, rebased)
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			err = os.MkdirAll(absPath, header.FileInfo().Mode())
+			if err != nil {
+				op.Errorf("Failed to create directory%s: %s", absPath, err)
+				return err
+			}
+		case tar.TypeSymlink:
+			err := os.Symlink(header.Linkname, absPath)
+			if err != nil {
+				op.Errorf("Failed to create symlink %s->%s: %s", absPath, header.Linkname, err)
+				return err
+			}
+		case tar.TypeReg:
+			f, err := os.OpenFile(absPath, fileWriteFlags, header.FileInfo().Mode())
+			if err != nil {
+				op.Errorf("Failed to open file %s: %s", absPath, err)
+				return err
+			}
+			_, err = io.Copy(f, tr)
+			// TODO: add ctx.Done cancellation
+			f.Close()
+			if err != nil {
+				return err
+			}
+		default:
+			// TODO: add support for special file types - otherwise we will do absurd things such as read infinitely from /dev/random
+		}
+		op.Debugf("Finished writing to: %s", absPath)
+	}
+	return nil
+}
+
+// Unpack hooks into a binary present in the appliance vm called unpack in order to execute InvokeUnpack inside of a chroot. This method works identically to InvokeUnpack, except that it will not function in areas where the binary is not present at /bin/unpack
+func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string) error {
+
+	fi, err := os.Stat(root)
+	if err != nil {
+		// the target unpack path does not exist. We should not get here.
+		op.Errorf("tar unpack target does not exist: %s", root)
+		return err
+	}
+
+	if !fi.IsDir() {
+		err := fmt.Errorf("unpack root target is not a directory: %s", root)
+		op.Error(err)
```

op.Errorf()

---

I want to return an error here, which is why I use `fmt.Errorf` to create a new error

---

Everywhere else, there is already an error to return.